File: programmers/level3/118670.js
function solution(rc, operations) {
    let answer = Array.from(Array(rc.length), () => Array(rc[0].length).fill(0));
    operations.forEach((operation) => {
        if (operation === 'Rotate') {
            for (let i = 0; i < rc.length; i++) {
                for (let j = 0; j < rc[0].length; j++) {
                    //오른쪽으로 이동
                    if (i === 0) {
                        if (j === 0) {
                            answer[i][j] = rc[i + 1][j];
                        } else {
                            answer[i][j] = rc[i][j - 1];
                        }
                    }
                    // 왼쪽으로 이동
                    else if (i === rc.length - 1) {
                        if (j === rc[0].length - 1) {
                            answer[i][j] = rc[i - 1][j];
                        } else {
                            answer[i][j] = rc[i][j + 1];
                        }
                    }
                    // 아래쪽으로 이동
                    else if (j === rc[0].length - 1) {
                        if (i > 0 && i < rc.length - 1) {
                            answer[i][j] = rc[i - 1][j];
                        }
                    }
                    // 위쪽으로 이동
                    else if (j === 0) {
                        if (i > 0 && i < rc.length - 1) {
                            answer[i][j] = rc[i + 1][j];
                        }
                    }
                }
            }
            for (let i = 0; i < rc.length; i++) {
                for (let j = 0; j < rc[0].length; j++) {
                    if (answer[i][j] === 0) {
                        answer[i][j] = rc[i][j];
                    }
                }
            }
        }

        if (operation === 'ShiftRow') {
            for (let i = 0; i < rc.length; i++) {
                if (i === 0) {
                    answer[i] = rc[rc.length - 1];
                } else {
                    answer[i] = rc[i - 1];
                }
            }
        }
        rc = answer.map((arr) => arr.slice());
    });
    return answer;
}

console.log(
    solution(
        [
            [1, 2, 3, 4],
            [5, 6, 7, 8],
            [9, 10, 11, 12],
            [13, 14, 15, 16],
            [17, 18, 19, 20],
        ],
        ['Rotate']
    )
);

File: programmers/level3/118670.py
def shift(rc,n,m):
    answer = [[0 for _ in range(m)] for _ in range(n)]

    for j in range(m):
        answer[0][j] = rc[n-1][j]
    for i in range(n-1):
        for j in range(m):
            answer[i+1][j] = rc[i][j]
    
    return answer

def rotate(rc,n,m):
    dx = [0,1,0,-1]
    dy = [1,0,-1,0]
    x,y = 0,1
    i = 0
    before = rc[0][0]
    while True:
        if x == 0 and y == 0:
            before,rc[x][y] = rc[x][y],before 
            break
        nx = x + dx[i]
        ny = y + dy[i]
        if 0 <= nx < n and 0 <= ny < m:
            before,rc[x][y] = rc[x][y],before 
            x = nx
            y = ny
        else:
            i += 1
    return rc


def solution(rc, operations):
    answer = rc
    for operation in operations:
        if operation == 'Rotate':
            answer = rotate(answer,len(rc),len(rc[0]))
        if operation == 'ShiftRow':
            answer = shift(answer,len(rc),len(rc[0]))
        print_rc(answer)
    return answer

def print_rc(rc):
    for i in rc:
        print(i)
    print("")

print(solution(	[[8, 6, 3], [3, 3, 7], [8, 4, 9]], ["Rotate", "ShiftRow", "ShiftRow"]))